### DISCUSSION TITLE
BlockSubdomainModifier?

### DISCUSSION THREAD

**Discussion by jmeier** [idaholab/moose#26952] [Q&A Modules: Solid mechanics]
Dear Community,

Following the procedure recommended [here](https://mooseframework.inl.gov/framework/contributing.html), I would like to propose a "BlockSubdomainModifier" for discussion. 

**Motivation and current situation**
With the group of "XYZSubdomainModifiers" Moose provides a tool to re-assign elements to different subdomains (blocks) during an simulation run. In tensor mechanis, this approach might be used to model the removal/excavation and refill. For geometrical complex models, the most efficient way defining the volumes to be excavated/refilled might be defined by means of subdomains. This leads to the task of moving whole blocks from one subdomain to another at pre-known points in time in a Moose-model. For this purpose the use of a [CoupledVarThresholdElementSubdomainModifier](https://mooseframework.inl.gov/source/userobjects/CoupledVarThresholdElementSubdomainModifier.html) is shown in #26941.

As shown in #26941, to just move one block once to another subdomain a Moose input file needs at least an AuxVariable, an FunctionAux, and a CoupledVarThresholdElementSubdomainModifier. Naïvely one could argue, it takes 36 lines of code to reassign a single block once to another subdomain - which is a source of error on its own. In geotechnical or mining applications it is necessary to reassign several blocks during one simulation.

**Proposed BlockSubdomainModifier**
To ease re-assignment, make Moose input files shorter, and reduce potential sources of error I'd like to propose an BlockSubdomainModifier. The BlockSubdomainModifier inherits from [ElementSubdomainModifier](https://github.com/idaholab/moose/blob/550348871a2dbd8cb284beddc5a18fead39780f7/framework/src/userobjects/ElementSubdomainModifier.C) (and therefore inheriting also the handling of the boundaries). The new UserObject takes a path to a CSV formates file containing the information on when to move which block where. Therefore, only one UserObject will be needed for all block reassignments in a given model.

Snipped of the input file using the proposed BlockSubdomainModifier:
```
[UserObjects]
  [BlockSubdomainModifier1]
    type = BlockSubdomainModifier
    file_name = 'stages-tab.csv'
    header = 'ON'
    execute_on = 'INITIAL TIMESTEP_BEGIN'
  []
[]
```

The corresponding CSV-File may contain a header and 3 columns (time, source-subdomain, and destination-subdomain):
```
time,   BlockFrom,  BlockTo
0.40,   Box1,       8
0.20,   Box2,       8
0.60,   8,          Box1
```

As can be seen in the CSV above, also subdomain-names are permissible (to improve readability and debugging). To allow for this, this proposal also includes the implementation of the DelimitedFileReaderOfString which is copied from the [DelimitedFileReader](https://github.com/idaholab/moose/blob/550348871a2dbd8cb284beddc5a18fead39780f7/framework/src/utils/DelimitedFileReader.C) (and removed all the stuff connected to conversion to double).

**Impact to existing Moose apps**
As this proposal only adds functionality and does not change any existing code, existing Moose apps should be able to cope with the proposed changes without any problems.

I would be delighted if you could consider this proposal. 

**Edit** Please find the code [here](https://github.com/grunerjmeier/moose/tree/BlockSubdomainModifier) including:

- [BlockSubdomainModifier.C](https://github.com/grunerjmeier/moose/blob/BlockSubdomainModifier/framework/src/userobjects/BlockSubdomainModifier.C)
- [DelimitedFileReaderOfString.C](https://github.com/grunerjmeier/moose/blob/BlockSubdomainModifier/framework/src/utils/DelimitedFileReaderOfString.C)

**Comment by GiudGiud** [comment]
Hello

We'll want @hugary1995 's opinion on this and possibly @lindsayad for awareness / re-direction to other stakeholders at least. 
Maybe @dschwen ?

These are good ideas overall and I think you have figured out a lot of the implementation already. 
One problem is the name. A block is the same as a subdomain so a BlockSubdomainModifier is not something that is easily understandable. 
Given the design I would say: `TimedSubdomainModifier`.
This conveys well that there are no other decision criteria on when to change the mesh. Also since there's "Element" in the name, it s understood that entire subdomains will change at once.

I can see simple input parameters as an alternative to the CSV file for simple problems. And for your large problem, use the CSV file option.

Guillaume

**Reply by lindsayad** [reply]
This seems reasonable to me

**Reply by jmeier** [reply]
Dear @GiudGiud,

I am happy about the positive feedback so far.

For the naming: 
Agreed, `TimedSubdomainModifier` describes the UserObject more meaningfully and better.

Simple input parameters as an alternative to the CSV:
I assume that the `TimedSubdomainModifier` should then behave in the same way as [PiecewiseConstant]( https://mooseframework.inl.gov/moose/source/functions/PiecewiseConstant.html) (having `xy_data` and `data_file`), for example? Not by means of a `TimedSubdomainModifier` (taking the data directly via parameter) and a `TimedSubdomainModifierCSV` using a CSV file?

Jörg

**Reply by GiudGiud** [reply]
Hello

Piecewise constant is an example of a dual input object yes.

I would prefer splitting the parameters for the input file parameters.

So 3 vector parameters
Old blocks
New blocks
Block_swap_time

It s more readable imo

Guillaume

**Reply by hugary1995** [reply]
This sounds like a great idea to me, and the proposed design seems reasonable to me.

Adding to @GiudGiud 's suggestion, I'd like to see part of the proposed capability moved into an intermediate class. So, instead of a concrete `TimedSubdomainModifier`, I suggest you create an intermediate abstract class called `TimedElementSubdomainModifier`, which doesn't perform the actual job. Add an additional input parameter `times` or `target_times` in `TimedElementSubdomainModifier`. All the logic related to target times should go into this intermediate class. For example, we should allow users to specify the target times with a vector, i.e. `target_times = '0.2 0.4 0.6'`. Once you have this intermediate class, let the existing `ThresholdElementSubdomainModifier` inherit from it, so that the new capability you add will also benefit all the existing modifiers.

The inheritance diagram I'm imagining (could be improved, for sure) is

```
                  ElementSubdomainModifier
                             |
                 TimedElementSubdomainModifier
                        |              |  
ThresholdElementSubdomainModifier    CSVSubdomainModifier
              |
CoupledVarThresholdElementSubdomainModifier
```

**Comment by jmeier** [comment]
Dear all,

Thank you, @GiudGiud, @lindsayad, and @hugary1995 for the positive feedback!

Following the suggestions of @GiudGiud and @hugary1995 I've created a new branch: [TimedSubdomainModifier](https://github.com/grunerjmeier/moose/tree/TimedSubdomainModifier)

The changes are:
- The user object is now called `TimedSubdomainModifier` [.h-file](https://github.com/grunerjmeier/moose/blob/TimedSubdomainModifier/framework/include/userobjects/TimedSubdomainModifier.h) [-c-file](https://github.com/grunerjmeier/moose/blob/TimedSubdomainModifier/framework/src/userobjects/TimedSubdomainModifier.C)
- The `TimedSubdomainModifier` now inherits of the newly created `TimedElementSubdomainModifier` [.h-file](https://github.com/grunerjmeier/moose/blob/TimedSubdomainModifier/framework/include/userobjects/TimedElementSubdomainModifier.h) [.c-file](https://github.com/grunerjmeier/moose/blob/TimedSubdomainModifier/framework/src/userobjects/TimedElementSubdomainModifier.C)
- The `TimedSubdomainModifier` now has properties to provide the times ans blocks directly in the input file -or- via file name using a CSV file. Please see below.
- The `TimedElementSubdomainModifier` does some work related to the timestamps (sorting them, providing `t_from_exclusive` and `t_to_inclusive` for the function `onComputeSubdomainID`
- The `DelimitedFileReaderOfString` remains unchanged [.h-file](https://github.com/grunerjmeier/moose/blob/TimedSubdomainModifier/framework/include/utils/DelimitedFileReaderOfString.h)  [.c-file](https://github.com/grunerjmeier/moose/blob/TimedSubdomainModifier/framework/src/utils/DelimitedFileReaderOfString.C)
- Blocks may be given via id or name (applies to the vector parameters and the CSV as well) to ease readability and model debugging.

With this changes the usage of the `TimedSubdomainModifier` looks like:

```
[UserObjects]
    [SubdomainModifier1]
      type = TimedSubdomainModifier
      times       = '0.4  0.2 '
      blocks_from = 'Box1 Box2'
      blocks_to   = '8    8   '
      execute_on = 'INITIAL TIMESTEP_BEGIN'
    []
[]
```

Or like this:
```
[UserObjects]
    [SubdomainModifier1]
      type = TimedSubdomainModifier
      data_file = 'stages-tab.csv'
      header = 'ON'
      execute_on = 'INITIAL TIMESTEP_BEGIN'
    []
[]
```

What do you think?

**Reply by GiudGiud** [reply]
Hello

user interface looks great. Will you make a PR to contribute your work to MOOSE? I can review the code then

Guillaume

**Reply by jmeier** [reply]
Dear @GiudGiud, 

Please find my PR here: #27092 

Kind regards,
Jörg
